# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MoHussein197/FlyRank-Intern/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

## 1. My lane and why

I choose **Lane 2: Refresh / Content Opportunity Scoring**.

**Why this lane?**
Content teams have limited bandwidth and cannot update thousands of articles at once.
Instead of guessing which articles need rewriting, this project builds a data-driven
prioritization queue. It identifies high-impression pages that are decaying or stale
so that human writers focus their time on pages where a refresh has the highest potential impact.

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

## 2. The question: decision, action, cost of a wrong call

* **Unit of Analysis:** A single published content item (page) for a given client (`content_id`).
* **Decision Improved:** Deciding which specific pages to assign to the editorial team for a content refresh or expansion this week.
* **Actor taking action:** SEO Specialists and Content Operations Managers.
* **Cost of a wrong call (False Positive vs. False Negative):**
  * *False Positive (flagging a healthy/stable page for refresh):* Wastes expensive editorial time and effort rewriting content that was already performing well.
  * *False Negative (missing a decaying high-demand page):* Results in continued loss of organic search traffic and potential revenue to competitors.
* **Why ML/Data helps:** ML can aggregate multi-dimensional signals (impressions, historical decay, position tiers, freshness) to rank thousands of pages objectively, beating simple 1-variable rules.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [4]:
import pandas as pd
import numpy as np

# Load the starter anonymized CSV
df = pd.read_csv("/content/content_refresh_anonymized.csv")

# Metric 1: Total pages vs Declining pages proportion
total_pages = len(df)
declining_pages = (df["trend_direction"].str.lower() == "down").sum()
decline_rate = (declining_pages / total_pages) * 100

# Metric 2: High visibility pages (impressions >= 500) that are decaying
high_vis_decay = df[(df["impressions_90d"] >= 500) & (df["trend_direction"].str.lower() == "down")]

# Metric 3: Stale pages (not updated in >= 180 days) with ongoing demand (impressions >= 500)
stale_high_demand = df[(df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 500)]

print(f"1. Total Pages Analyzed: {total_pages:,} | Overall Decline Rate: {decline_rate:.1f}%")
print(f"2. High-Visibility Decaying Pages (Impressions >= 500 & Trend == 'down'): {len(high_vis_decay):,}")
print(f"3. Stale High-Demand Pages (Unupdated >= 180 days & Impressions >= 500): {len(stale_high_demand):,}")

1. Total Pages Analyzed: 30,000 | Overall Decline Rate: 54.2%
2. High-Visibility Decaying Pages (Impressions >= 500 & Trend == 'down'): 9,961
3. Stale High-Demand Pages (Unupdated >= 180 days & Impressions >= 500): 17


### Key Insights from the Starter Dataset:
1. **Widespread Traffic Decay:** Out of 30,000 total pages analyzed, **54.2%** are currently experiencing a downward traffic trend (`trend_direction == 'down'`).
2. **Massive High-Value Opportunity:** There are **9,961 high-visibility pages** (generating ≥ 500 impressions in 90 days) that are actively declining. This represents a huge, high-priority pool where prioritization and refresh recommendations can prevent massive traffic loss.
3. **Immediate Low-Hanging Fruit:** **17 pages** have active demand (≥ 500 impressions) despite not being updated for over 6 months (≥ 180 days), making them clear candidates for quick-win content updates.

These numbers confirm that traffic decay is widespread across the inventory and that an ML-based ranking queue is required to prioritize which of the 9,961 declining pages the editorial team should review first.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

## 4. Careful words: what I can and can't claim

* **What I CAN claim:**
  * Observed patterns and associations between page characteristics (e.g., freshness, position tier) and traffic movement.
  * Directional risk scores and ranked decision-support queues to optimize human editorial review.
  * Measured performance improvements of learned models over baseline heuristic rules using holdout cross-validation.

* **What I CANNOT claim:**
  * Causal proof that updating a page will guarantee a recovery in search traffic.
  * "Reverse-engineering Google's search algorithm" or predicting ranking factors with deterministic certainty.
  * Exact claims about client names, domain strategies, or raw private search queries (due to pseudonymization and observational study limits).

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [X] Every section above is filled — markdown thinking AND the code that backs it
- [X] The notebook runs top to bottom with no errors (Runtime → Run all)
- [X] No client names, URLs, or private queries anywhere
- [X] My claims use careful words: observed, measured, directional, decision-support
- [X] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.